File: src/lib/tauri/sftp.ts
import { Channel, invoke } from "@tauri-apps/api/core";

import type {
  AttachRemoteFileTransfersInput,
  DownloadRemoteFileInput,
  LocalDownloadTargetSelection,
  LocalUploadFileSelection,
  RemoteDirectory,
  RemoteDownloadResult,
  RemoteFileTransferControlStatus,
  RemoteFileTransferProgress,
  RemoteUploadResult,
  SelectLocalDownloadTargetInput,
  SelectLocalUploadFilesInput,
  UploadRemoteFileInput,
} from "@/types/sftp";

const LIST_REMOTE_DIRECTORY_COMMAND = "list_remote_directory";
const CREATE_REMOTE_DIRECTORY_COMMAND = "create_remote_directory";
const CREATE_REMOTE_FILE_COMMAND = "create_remote_file";
const RENAME_REMOTE_ENTRY_COMMAND = "rename_remote_entry";
const DELETE_REMOTE_ENTRY_COMMAND = "delete_remote_entry";
const SELECT_LOCAL_UPLOAD_FILES_COMMAND = "select_local_upload_files";
const SELECT_LOCAL_DOWNLOAD_TARGET_COMMAND = "select_local_download_target";
const ATTACH_REMOTE_FILE_TRANSFERS_COMMAND = "attach_remote_file_transfers";
const UPLOAD_REMOTE_FILE_COMMAND = "upload_remote_file";
const DOWNLOAD_REMOTE_FILE_COMMAND = "download_remote_file";
const PAUSE_REMOTE_FILE_TRANSFER_COMMAND = "pause_remote_file_transfer";
const CANCEL_REMOTE_FILE_TRANSFER_COMMAND = "cancel_remote_file_transfer";

export function listRemoteDirectory(sessionId: string, path?: string) {
  return invoke<RemoteDirectory>(LIST_REMOTE_DIRECTORY_COMMAND, {
    sessionId,
    path,
  });
}

export function createRemoteDirectory(
  sessionId: string,
  parentPath: string,
  name: string,
) {
  return invoke<string>(CREATE_REMOTE_DIRECTORY_COMMAND, {
    sessionId,
    parentPath,
    name,
  });
}

export function createRemoteFile(sessionId: string, parentPath: string, name: string) {
  return invoke<string>(CREATE_REMOTE_FILE_COMMAND, {
    sessionId,
    parentPath,
    name,
  });
}

export function renameRemoteEntry(sessionId: string, path: string, newName: string) {
  return invoke<string>(RENAME_REMOTE_ENTRY_COMMAND, {
    sessionId,
    path,
    newName,
  });
}

export function deleteRemoteEntry(sessionId: string, path: string) {
  return invoke<void>(DELETE_REMOTE_ENTRY_COMMAND, { sessionId, path });
}

export function selectLocalUploadFiles(input: SelectLocalUploadFilesInput) {
  return invoke<LocalUploadFileSelection[]>(SELECT_LOCAL_UPLOAD_FILES_COMMAND, {
    input,
  });
}

export function selectLocalDownloadTarget(input: SelectLocalDownloadTargetInput) {
  return invoke<LocalDownloadTargetSelection | null>(
    SELECT_LOCAL_DOWNLOAD_TARGET_COMMAND,
    { input },
  );
}

export function attachRemoteFileTransfers(input: AttachRemoteFileTransfersInput) {
  return invoke<void>(ATTACH_REMOTE_FILE_TRANSFERS_COMMAND, { input });
}

export function uploadRemoteFile(
  input: UploadRemoteFileInput,
  onProgress: (progress: RemoteFileTransferProgress) => void,
) {
  const progressChannel = new Channel<RemoteFileTransferProgress>(onProgress);
  return invoke<RemoteUploadResult>(UPLOAD_REMOTE_FILE_COMMAND, {
    input,
    onProgress: progressChannel,
  });
}

export function downloadRemoteFile(
  input: DownloadRemoteFileInput,
  onProgress: (progress: RemoteFileTransferProgress) => void,
) {
  const progressChannel = new Channel<RemoteFileTransferProgress>(onProgress);
  return invoke<RemoteDownloadResult>(DOWNLOAD_REMOTE_FILE_COMMAND, {
    input,
    onProgress: progressChannel,
  });
}

export function cancelRemoteFileTransfer(transferId: string) {
  return invoke<RemoteFileTransferControlStatus>(CANCEL_REMOTE_FILE_TRANSFER_COMMAND, {
    transferId,
  });
}

export function pauseRemoteFileTransfer(transferId: string) {
  return invoke<RemoteFileTransferControlStatus>(PAUSE_REMOTE_FILE_TRANSFER_COMMAND, {
    transferId,
  });
}

File: src-tauri/src/commands/sftp.rs
use tauri::ipc::Channel;
use tauri::{State, WebviewWindow};
use tokio::sync::mpsc;

use crate::managers::sessions::SshSessionManager;
use crate::models::error::CommandError;
use crate::models::sftp::{
    AttachRemoteFileTransfersInput, DownloadRemoteFileInput, LocalDownloadTargetSelectionDto,
    LocalUploadFileSelectionDto, RemoteDirectoryDto, RemoteDownloadResultDto,
    RemoteFileTransferCancelStatusDto, RemoteFileTransferProgressDto, RemoteUploadResultDto,
    SelectLocalDownloadTargetInput, SelectLocalUploadFilesInput, UploadRemoteFileInput,
};

const TRANSFER_EVENT_QUEUE_CAPACITY: usize = 32;

#[tauri::command]
pub async fn list_remote_directory(
    session_id: String,
    path: Option<String>,
    sessions: State<'_, SshSessionManager>,
) -> Result<RemoteDirectoryDto, CommandError> {
    sessions
        .list_remote_directory(&session_id, path.as_deref())
        .await
        .map(Into::into)
        .map_err(Into::into)
}

#[tauri::command]
pub async fn create_remote_directory(
    session_id: String,
    parent_path: String,
    name: String,
    sessions: State<'_, SshSessionManager>,
) -> Result<String, CommandError> {
    sessions
        .create_remote_directory(&session_id, &parent_path, &name)
        .await
        .map_err(Into::into)
}

#[tauri::command]
pub async fn create_remote_file(
    session_id: String,
    parent_path: String,
    name: String,
    sessions: State<'_, SshSessionManager>,
) -> Result<String, CommandError> {
    sessions
        .create_remote_file(&session_id, &parent_path, &name)
        .await
        .map_err(Into::into)
}

#[tauri::command]
pub async fn rename_remote_entry(
    session_id: String,
    path: String,
    new_name: String,
    sessions: State<'_, SshSessionManager>,
) -> Result<String, CommandError> {
    sessions
        .rename_remote_entry(&session_id, &path, &new_name)
        .await
        .map_err(Into::into)
}

#[tauri::command]
pub async fn delete_remote_entry(
    session_id: String,
    path: String,
    sessions: State<'_, SshSessionManager>,
) -> Result<(), CommandError> {
    sessions
        .delete_remote_entry(&session_id, &path)
        .await
        .map_err(Into::into)
}

#[tauri::command]
pub async fn select_local_upload_files(
    input: SelectLocalUploadFilesInput,
    window: WebviewWindow,
    sessions: State<'_, SshSessionManager>,
) -> Result<Vec<LocalUploadFileSelectionDto>, CommandError> {
    sessions
        .select_local_upload_files(&window, &input.session_id, &input.remote_directory)
        .await
        .map(|selections| selections.into_iter().map(Into::into).collect())
        .map_err(Into::into)
}

#[tauri::command]
pub async fn select_local_download_target(
    input: SelectLocalDownloadTargetInput,
    window: WebviewWindow,
    sessions: State<'_, SshSessionManager>,
) -> Result<Option<LocalDownloadTargetSelectionDto>, CommandError> {
    sessions
        .select_local_download_target(
            &window,
            &input.session_id,
            &input.remote_path,
            &input.default_file_name,
        )
        .await
        .map(|selection| selection.map(Into::into))
        .map_err(Into::into)
}

#[tauri::command]
pub async fn attach_remote_file_transfers(
    input: AttachRemoteFileTransfersInput,
    sessions: State<'_, SshSessionManager>,
) -> Result<(), CommandError> {
    sessions
        .attach_remote_file_transfers(&input.transfer_ids)
        .await
        .map_err(Into::into)
}

#[tauri::command]
pub async fn upload_remote_file(
    input: UploadRemoteFileInput,
    on_progress: Channel<RemoteFileTransferProgressDto>,
    sessions: State<'_, SshSessionManager>,
) -> Result<RemoteUploadResultDto, CommandError> {
    let progress_sender = forward_transfer_progress(on_progress);

    sessions
        .upload_remote_file(input.transfer_id, progress_sender)
        .await
        .map(Into::into)
        .map_err(Into::into)
}

#[tauri::command]
pub async fn download_remote_file(
    input: DownloadRemoteFileInput,
    on_progress: Channel<RemoteFileTransferProgressDto>,
    sessions: State<'_, SshSessionManager>,
) -> Result<RemoteDownloadResultDto, CommandError> {
    let progress_sender = forward_transfer_progress(on_progress);

    sessions
        .download_remote_file(input.transfer_id, progress_sender)
        .await
        .map(Into::into)
        .map_err(Into::into)
}

#[tauri::command]
pub async fn cancel_remote_file_transfer(
    transfer_id: String,
    sessions: State<'_, SshSessionManager>,
) -> Result<RemoteFileTransferCancelStatusDto, CommandError> {
    sessions
        .cancel_remote_file_transfer(&transfer_id)
        .await
        .map(Into::into)
        .map_err(Into::into)
}

fn forward_transfer_progress(
    on_progress: Channel<RemoteFileTransferProgressDto>,
) -> mpsc::Sender<crate::domain::sftp::RemoteFileTransferProgress> {
    let (progress_sender, mut progress_receiver) = mpsc::channel::<
        crate::domain::sftp::RemoteFileTransferProgress,
    >(TRANSFER_EVENT_QUEUE_CAPACITY);
    tauri::async_runtime::spawn(async move {
        while let Some(progress) = progress_receiver.recv().await {
            if on_progress.send(progress.into()).is_err() {
                break;
            }
        }
    });
    progress_sender
}
